osu.Game/Screens/Edit/Changes/PropertyChange.cs

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+namespace osu.Game.Screens.Edit.Changes
+{
+    /// <summary>
+    /// Represents a single property update on a given <see cref="Target"/>.
+    /// </summary>
+    /// <typeparam name="TTarget">Type of the object owning the property</typeparam>
+    /// <typeparam name="TValue">Type of the property to update</typeparam>
+    public abstract class PropertyChange<TTarget, TValue> : IRevertibleChange where TTarget : class
```

Comment:
I still have a fair bit of skepticism towards this thing still. Generic usage will always have me stop and think "is it *really* necessary though?"

I recall you saying something that this "helps alleviate conflicts" when talking about collaborative editing or something. Can you elaborate on this? Because to my naive mind, I'm not convinced granularity at property level is required. I'd say the 'unit of operation' in the editor is a single hitobject, timing point, etc.

The point I'm trying to make is that it seems this is trying to make flows like "user A moves slider, user B changes its path" work without incurring conflicts - and I guess what I'm trying to say to that is I'm not sure that sort of thing would make sense? Those two edits are something that *should* conflict to me.

Reply:
Those two edits being conflicting makes sense, however it's more about how you deal with the conflict. That particular case, while probably being a bit jank if it happens, shouldn't really break anything. Dealing with conflicts on a per-property level allows just doing a simple last-write-wins appraoch for everything. I think the moment you introduce custom handling for every edge case like this complexity is just gonna skyrocket.

And what would the conflict resolution for this case even look like? Whoever touched the slider last will have their drag operation canceled? Everyone sees their local changes until they're done dragging and then it snaps to whichever one finished last? Will all changes to the same object be considered a conflict? What if someone selects all objects & i.e. modifies their hitsound, will that create conflicts with someone moving an object at that moment?
Those updates can of course be done on a per-object level too with i.e. a `UpdateHitObjectCommand` like I'm doing in osucad but they should still be able to pick which properties should be updated & I wanted to avoid that to have stuff like this
```ts
// its fine to do stuff like this in js since it's all dynamic anyways but not so sure about c#
function applyPatch(circle: HitCircle, patch: Partial<SerializedHitCircle>) {
  for (const key in patch) {
    let value = patch[key] as any
    if (key in circlePropertyDeserializers)
      value = circlePropertyDeserializers[key](value)
    
    circle[key] = value
  }
}
```
or this
```ts
function applyPatch(circle: HitCircle, patch: Partial<SerializedHitCircle>) {
  if (patch.startTime !== undefined)
    circle.startTime = patch.startTime;
  if (patch.position !== undefined)
    circle.position = Vec2.from(patch.position);
  if (patch.comboOffset !== undefined)
    circle.comboOffset = patch.comboOffset;
  if (patch.newCombo !== undefined)
    circle.newCombo = patch.newCombo;
  if (patch.hitSound !== undefined)
    circle.hitSound = deserializeHitSound(patch.hitSound);
}
```